File: mergeMeetings.js


function mergeRanges(meetings) {
  //define variables
  let copiedMeetings = meetings.slice(0);
  let sortedMeetings = copiedMeetings.sort((a, b) => (a.startTime > b.startTime) ? 1 : -1)
  let finalRanges = [sortedMeetings[0]];

  for (let i = 1; i <= sortedMeetings.length - 1; i++) {
    
    // current meeting is current object on sorted meetings array 
    let currentMeeting = sortedMeetings[i];
    // last meeting set refers to last object aka meeting on the final array 
    let lastMeetingSet = finalRanges[finalRanges.length - 1]; // looks at last item in final ranges array 

    if ( currentMeeting && (currentMeeting.startTime <= lastMeetingSet.endTime)) {
      // check end times and merge with existing meeting object
      lastMeetingSet.endTime = Math.max(currentMeeting.endTime, lastMeetingSet.endTime);
    } else {
      // otherwise push the currentMeeting onto finalRanges
      finalRanges.push(currentMeeting);
    }
  }
  
  return finalRanges; 
}


/* --------------------------------------------------- Tests ------------------------------------------------*/

// let desc = 'meetings overlap';
// let actual = mergeRanges([{ startTime: 1, endTime: 3 }, { startTime: 2, endTime: 4 }]);
// let expected = [{ startTime: 1, endTime: 4 }];
// assertArrayEquals(actual, expected, desc);

// desc = 'meetings touch';
// actual = mergeRanges([{ startTime: 5, endTime: 6 }, { startTime: 6, endTime: 8 }]);
// expected = [{ startTime: 5, endTime: 8 }];
// assertArrayEquals(actual, expected, desc);

// desc = 'meeting contains other meeting';
// actual = mergeRanges([{ startTime: 1, endTime: 8 }, { startTime: 2, endTime: 5 }]);
// expected = [{ startTime: 1, endTime: 8 }];
// assertArrayEquals(actual, expected, desc);

// desc = 'meetings stay separate';
// actual = mergeRanges([{ startTime: 1, endTime: 3 }, { startTime: 4, endTime: 8 }]);
// expected = [{ startTime: 1, endTime: 3 }, { startTime: 4, endTime: 8 }];
// assertArrayEquals(actual, expected, desc);

// desc = 'multiple merged meetings';
// actual = mergeRanges([
//   { startTime: 1, endTime: 4 },
//   { startTime: 2, endTime: 5 },
//   { startTime: 5, endTime: 8 },
// ]);
// expected = [{ startTime: 1, endTime: 8 }];
// assertArrayEquals(actual, expected, desc);

// desc = 'meetings not sorted';
// actual = mergeRanges([
//   { startTime: 5, endTime: 8 },
//   { startTime: 1, endTime: 4 },
//   { startTime: 6, endTime: 8 },
// ]);
// expected = [{ startTime: 1, endTime: 4 }, { startTime: 5, endTime: 8 }];
// assertArrayEquals(actual, expected, desc);

// desc = 'oneLongMeetingContainsSmallerMeetings';
// actual = mergeRanges([
//   { startTime: 1, endTime: 10 },
//   { startTime: 2, endTime: 5 },
//   { startTime: 6, endTime: 8 },
//   { startTime: 9, endTime: 10 },
//   { startTime: 10, endTime: 12 }
// ]);
// expected = [
//   { startTime: 1, endTime: 12 }
// ];
// assertArrayEquals(actual, expected, desc);

// desc = 'sample input';
// actual = mergeRanges([
//   { startTime: 0, endTime: 1 },
//   { startTime: 3, endTime: 5 },
//   { startTime: 4, endTime: 8 },
//   { startTime: 10, endTime: 12 },
//   { startTime: 9, endTime: 10 },
// ]);
// expected = [
//   { startTime: 0, endTime: 1 },
//   { startTime: 3, endTime: 8 },
//   { startTime: 9, endTime: 12 },
// ];
// // assertArrayEquals(actual, expected, desc);

// function assertArrayEquals(a, b, desc) {
//   // Sort the keys in each meeting to avoid
//   // failing based on differences in key order.
//   orderedA = a.map( function(meeting) {
//     return JSON.stringify(meeting, Object.keys(meeting).sort());
//   });
//   orderedB = b.map( function(meeting) {
//     return JSON.stringify(meeting, Object.keys(meeting).sort());
//   });
//   const arrayA = JSON.stringify(orderedA);
//   const arrayB = JSON.stringify(orderedB);
//   if (arrayA !== arrayB) {
//     console.log(`${desc} ... FAIL: ${JSON.stringify(a)} != ${JSON.stringify(b)}`)
//   } else {
//     console.log(`${desc} ... PASS`);
//   }
// }

File: mergeMeetings.py
import unittest


# Merge meeting ranges

# O: a list of tuples representing occupied blocks of time (start, end)
# I: a list of tuples
# C: time ranges represented by integers (could be very large integers)
#    ranges not sorted
#    shouldn't mutate input
# E: a time range that could fit inside another time range
#    two ranges that touch but do not overlap
#    zero-length range
#
# Approach:
#   copy input to avoid mutating it
#   Sort meetings by ascending order of the first element in the tuple
#   we will build up our condensed ranges one by one
#   for each meeting in meetings,
#       if it starts before or at the end of the current range
#           update the end of our range if it is less than the end of the meeting
#       otherwise, add our current condensed range to the return list and base our next range on the current meeting


# meetings should be a list of tuples, where each tuple represents a time range (start, end)
# (2, 3)  # Meeting from 10:00 - 10:30 am
# (6, 9)  # Meeting from 12:00 - 1:30 pm
# meetings might be something like [(0, 1), (3, 5), (4, 8), (10, 12), (9, 10)]
#
# returns a similarly formatted list of tuples representing all continuously occupied blocks of time
def merge_ranges(meetings):
    meetings_copy = sorted(meetings)
    condensed_ranges = []

    range_start, range_end = meetings_copy[0]

    for meeting_start, meeting_end in meetings_copy:
        if meeting_start <= range_end:
            range_end = max(range_end, meeting_end)
        else:
            condensed_ranges.append((range_start, range_end))
            range_start = meeting_start
            range_end = meeting_end

    condensed_ranges.append((range_start, range_end))

    return condensed_ranges


# # Tests
#
# class Test(unittest.TestCase):
#
#     def test_meetings_overlap(self):
#         actual = merge_ranges([(1, 3), (2, 4)])
#         expected = [(1, 4)]
#         self.assertEqual(actual, expected)
#
#     def test_meetings_touch(self):
#         actual = merge_ranges([(5, 6), (6, 8)])
#         expected = [(5, 8)]
#         self.assertEqual(actual, expected)
#
#     def test_meeting_contains_other_meeting(self):
#         actual = merge_ranges([(1, 8), (2, 5)])
#         expected = [(1, 8)]
#         self.assertEqual(actual, expected)
#
#     def test_meetings_stay_separate(self):
#         actual = merge_ranges([(1, 3), (4, 8)])
#         expected = [(1, 3), (4, 8)]
#         self.assertEqual(actual, expected)
#
#     def test_multiple_merged_meetings(self):
#         actual = merge_ranges([(1, 4), (2, 5), (5, 8)])
#         expected = [(1, 8)]
#         self.assertEqual(actual, expected)
#
#     def test_meetings_not_sorted(self):
#         actual = merge_ranges([(5, 8), (1, 4), (6, 8)])
#         expected = [(1, 4), (5, 8)]
#         self.assertEqual(actual, expected)
#
#     def test_one_long_meeting_contains_smaller_meetings(self):
#         actual = merge_ranges([(1, 10), (2, 5), (6, 8), (9, 10), (10, 12)])
#         expected = [(1, 12)]
#         self.assertEqual(actual, expected)
#
#     def test_sample_input(self):
#         actual = merge_ranges([(0, 1), (3, 5), (4, 8), (10, 12), (9, 10)])
#         expected = [(0, 1), (3, 8), (9, 12)]
#         self.assertEqual(actual, expected)
#
#
# unittest.main(verbosity=2)